File: app/src/main/java/sk/styk/martin/apkanalyzer/util/file/FileUtils.java
package sk.styk.martin.apkanalyzer.util.file;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.PrintWriter;

/**
 * @author Martin Styk
 * @version 03.01.2018.
 */

public class FileUtils {

    public static void copy(File src, File dst) throws IOException {
        InputStream in = null;
        OutputStream out = null;
        try {
            in = new FileInputStream(src);
            out = new FileOutputStream(dst);

            byte[] buf = new byte[1024];
            int len;
            while ((len = in.read(buf)) > 0) {
                out.write(buf, 0, len);
            }
        } finally {
            if (in != null)
                in.close();
            if (out != null)
                out.close();
        }
    }

    public static void writeString(String content, String targetFilePath) throws IOException {
        PrintWriter printWriter = null;

        try {
            printWriter = new PrintWriter(targetFilePath);
            printWriter.print(content);
        } finally {
            if (printWriter != null)
                printWriter.close();
        }

    }
}

File: app/src/main/java/sk/styk/martin/apkanalyzer/util/file/FileUtils.kt
package sk.styk.martin.apkanalyzer.util.file

import android.os.Environment
import androidx.annotation.WorkerThread
import kotlinx.coroutines.delay
import kotlinx.coroutines.yield
import java.io.*

object FileUtils {

    val externalRoot by lazy { Environment.getExternalStorageDirectory() }
    val externalAppsDirectory by lazy { File(externalRoot, "ApkAnalyzer/") }

    fun toRelativePath(absolutePath: String) = absolutePath.substring(externalRoot.absolutePath.length + 1)

    @WorkerThread
    @Throws(IOException::class)
    suspend fun copy(src: InputStream, dst: File) {

        FileOutputStream(dst).use { output ->

            val buffer = ByteArray(10240)
            var len: Int = src.read(buffer)
            var readBytes = len
            while (len > 0) {
                yield()
                output.write(buffer, 0, len)
                len = src.read(buffer)
                readBytes += len
            }
        }
        delay(500)
    }

    @WorkerThread
    @Throws(IOException::class)
    fun writeString(content: String, targetFilePath: String) {

        PrintWriter(targetFilePath).use {
            it.print(content)
        }

    }


}
